### PR TITLE
feat: New validation hook to check if scale target is already managed

### PR DESCRIPTION
**Pull request by JorTurFer**
Signed-off-by: Jorge Turrado <jorge_turrado@hotmail.es>

**Note**: This PR uses my own fork of opa/cert-controller until the merge/reject [my PR](https://github.com/open-policy-agent/cert-controller/pull/52), If they don't merge it, I plan to fork their repo in kedacore organization with the changes

**Note 2**: I have removed duplicated test code for influx scaler because that test fails after adding the webhooks (due to multiple SO for the same workload)

### Checklist

- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)* https://github.com/kedacore/charts/pull/352 **Can't be done before the image is available because the checks fail**
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)* https://github.com/kedacore/keda-docs/pull/1027
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Relates to https://github.com/kedacore/keda/issues/3755
Relates to https://github.com/kedacore/keda/issues/3087
Relates to https://github.com/kedacore/keda/issues/3083
Relates to https://github.com/kedacore/keda/issues/3088

### CODE REVIEW

**Comment by JorTurFer** [comment]
> Semgrep found **1** [`run-as-non-root`](https://semgrep.dev/r/yaml.kubernetes.security.run-as-non-root.run-as-non-root) finding:
> 
> * config/webhooks/webhooks.yaml: [L29](https://github.com/kedacore/keda/blob/31d6a5c8323295fe5d63975bde5fe3bf4515966b/config/webhooks/webhooks.yaml#L29)
> 
> Container allows for running applications as root. This can result in privilege escalation attacks. Add 'runAsNonRoot: true' in 'securityContext' to prevent this.
> 
> Created by [run-as-non-root](https://semgrep.dev/r/yaml.kubernetes.security.run-as-non-root.run-as-non-root?utm_campaign=finding_notification&utm_medium=review_comment&utm_source=github&utm_content=rule).

solved

File: apis/keda/v1alpha1/scaledobject_webhook.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		if hpatargetAPI == incomingSotargetAPI &&
+			hpaTargetKind == incomingSoTargetKind &&
+			hpaTarget.Name == incomingSoTarget.Name {
+			owned := false
+			ownerName := ""
+			for _, owner := range hpa.OwnerReferences {
+				if owner.Kind == incomingSo.Kind {
+					ownerName = owner.Name
+					if owner.Name == incomingSo.Name {
+						owned = true
```
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_webhook.go:121]:
```suggestion
						owned = true
						break
```

File: apis/keda/v1alpha1/scaledobject_webhook.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+					}
+				}
+			}
+
+			if !owned {
+				var err error
+				if len(hpa.OwnerReferences) == 0 {
+					err = fmt.Errorf("the workload '%s' of type '%s/%s' is already managed by the hpa '%s'", incomingSoTarget.Name, incomingSotargetAPI, incomingSoTargetKind, hpa.Name)
+				} else {
+					err = fmt.Errorf("the workload '%s' of type '%s/%s' is already managed by the ScaledObject '%s'", incomingSoTarget.Name, incomingSotargetAPI, incomingSoTargetKind, ownerName)
```
Comment (zroubalik) [inline apis/keda/v1alpha1/scaledobject_webhook.go:131]: What if the HPA is owned by some other resource (not just ScaledObject)? This should message should be imho generic and api/kind/name should be parametrized.

File: apis/keda/v1alpha1/withtriggers_types.go [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KEDA Authors
+Copyright 2023 The KEDA Authors
```
Comment (zroubalik) [inline apis/keda/v1alpha1/withtriggers_types.go:2]: unrelevant change, pls remove

File: cmd/adapter/main.go [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KEDA Authors
+Copyright 2023 The KEDA Authors
```
Comment (zroubalik) [inline cmd/adapter/main.go:2]: unrelevant change, pls remove

File: cmd/operator/main.go [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KEDA Authors
+Copyright 2023 The KEDA Authors
```
Comment (zroubalik) [inline cmd/operator/main.go:2]: unrelevant change, pls remove

File: cmd/webhooks/main.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		CertDir:                webhookCertDir,
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start webhooks")
+		os.Exit(1)
+	}
+
+	// Make sure certs are generated and valid if cert rotation is enabled.
+	setupFinished := make(chan struct{})
+	if !disableCertRotation {
```
Comment (zroubalik) [inline cmd/webhooks/main.go:111]: what if user would like to add own certs?

File: controllers/keda/hpa.go [hunk 7]
```diff
@@ -29,7 +29,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
-	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	"github.com/kedacore/keda/v2/apis/keda/v1alpha1"
```
Comment (zroubalik) [inline controllers/keda/hpa.go:32]: why this change? TBH I prefer the original, so we have consistent naming of this dependency across the codebase.

File: controllers/keda/scaledobject_controller.go [hunk 8]
```diff
@@ -41,7 +41,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
-	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	"github.com/kedacore/keda/v2/apis/keda/v1alpha1"
```
Comment (zroubalik) [inline controllers/keda/scaledobject_controller.go:44]: same as above

File: pkg/provider/provider.go [hunk 9]
```diff
@@ -31,7 +31,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/custom-metrics-apiserver/pkg/provider"
 
-	kedav1alpha1 "github.com/kedacore/keda/v2/apis/keda/v1alpha1"
+	"github.com/kedacore/keda/v2/apis/keda/v1alpha1"
```
Comment (zroubalik) [inline pkg/provider/provider.go:34]: same as above

File: pkg/scaling/resolver/scale_resolvers.go [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KEDA Authors
+Copyright 2023 The KEDA Authors
```
Comment (zroubalik) [inline pkg/scaling/resolver/scale_resolvers.go:2]: unrelevant change, pls remove

File: pkg/util/env_resolver.go [hunk 11]
```diff
@@ -77,6 +77,15 @@ func GetClusterObjectNamespace() (string, error) {
 	return strData, nil
 }
 
+// GetPodNamespace returns the namespace for the pod
+func GetPodNamespace() string {
```
Comment (zroubalik) [inline pkg/util/env_resolver.go:81]: The same functionality is implemented in `adapter/main.go`, that one should use this common funcion as well.

File: config/rbac/role.yaml [hunk 12]
```diff
@@ -99,3 +109,23 @@ rules:
   - triggerauthentications/status
   verbs:
   - '*'
+---
```
Comment (zroubalik) [inline config/rbac/role.yaml:112]: why do we have this new, separate, role?

File: PROJECT [hunk 13]
```diff
@@ -12,16 +12,19 @@ resources:
   domain: keda.sh
   group: keda
   kind: ScaledObject
-  path: github.com/kedacore/keda/v2/apis/keda/v1alpha1
+  path: github.com/kedacore/keda/apis/keda/v1alpha1
```
[the remark this answers is not in the document]
Comment (JorTurFer) [reply PROJECT:15]: The problem was that operator-framework/kubebuilder didn't detect the APIs and i couldn't create the webhook. Once I changed this, the tooling worked again, but honestly, I didn't understand way and maybe it was because I was doing something wrong, if you think that v2 is worth, I can revert this change

[the remark this answers is not in the document]
Comment (zroubalik) [reply PROJECT:15]: ah okay, so keep it this way then!

File: apis/keda/v1alpha1/scaledobject_types.go [hunk 14]
```diff
@@ -184,6 +186,6 @@ func init() {
 }
 
 // GenerateIdentifier returns identifier for the object in for "kind.namespace.name"
-func (s *ScaledObject) GenerateIdentifier() string {
-	return GenerateIdentifier("ScaledObject", s.Namespace, s.Name)
+func (so *ScaledObject) GenerateIdentifier() string {
```
[the remark this answers is not in the document]
Comment (JorTurFer) [reply apis/keda/v1alpha1/scaledobject_types.go:189]: This change is required here (or in the new code), because the linter/rules requires that all the functions from the same struct has the same name for the item, if I name it here with `s`, I need to use `s` for every function. I can revert his change here and update the code in the other places to use `s` instead of `so` but I felt `so` more accurate with the element we are managing. 
I'm happy with both so as you prefer

[the remark this answers is not in the document]
Comment (zroubalik) [reply apis/keda/v1alpha1/scaledobject_types.go:189]: okay, let's keep it this way :)

File: apis/keda/v1alpha1/scaledobject_webhook.go (hunk 2 above)
Comment (JorTurFer) [reply apis/keda/v1alpha1/scaledobject_webhook.go:131]: This code is old and I forgot removing it. The first approach was to check all in the same function, but now we are checking 1st the SOs conflicts and then the HPA conflicts, so this else should never happen. If the code reach here, only the HPA case can happen

File: config/rbac/role.yaml (hunk 12 above)
Comment (JorTurFer) [reply config/rbac/role.yaml:112]: Do you mean in other file? kubebuilder has generated it here, let me check if I can split them

File: pkg/util/env_resolver.go (hunk 11 above)
Comment (JorTurFer) [reply pkg/util/env_resolver.go:81]: nice catch! updated :)

File: apis/keda/v1alpha1/scaledobject_webhook.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+	prommetrics "github.com/kedacore/keda/v2/pkg/prommetrics/webhook"
+)
+
+var scaledobjectlog = logf.Log.WithName("scaledobject-validation-webhook")
+
+var kc client.Client
+
+const (
+	defaultAPI  = "apps/v1"
```
[the remark this answers is not in the document]
Comment (JorTurFer) [reply apis/keda/v1alpha1/scaledobject_webhook.go:39]: I think that I can use [gvkr](https://github.com/kedacore/keda/blob/main/pkg/util/gvkr.go) (the mechanism for other places) and it solves the problem :)

File: cmd/webhooks/main.go (hunk 6 above)
Comment (JorTurFer) [reply cmd/webhooks/main.go:117]: They can do it. I'll update docs explaining how to do it, but it's just setting the argument `disable-cert-rotation: true` and placing the certificates in the correct path (`/certs` as default but configurable with the argument `webhook-cert-dir`)

File: config/rbac/role.yaml (hunk 12 above)
Comment (JorTurFer) [reply config/rbac/role.yaml:112]: I have checked and the only way I have found is splitting the rbac generation for each service, setting which files should be used for each service instead of using `./...` . do you know other way to do it?

File: BUILD.md [hunk 16]
```diff
@@ -223,6 +223,64 @@ You can query list metrics executing `curl --insecure https://localhost:6443/api
 
 If you prefer to use an authenticated user, you can use a user or service account with access over external metrics API adding their token as authorization header in `curl`, ie: `curl -H "Authorization:Bearer TOKEN" --insecure https://localhost:6443/apis/external.metrics.k8s.io/v1beta1/`
 
+### Webhooks
```
Comment (tomkerkhove) [inline BUILD.md:226]: See feedback on docs but I'd remove all "webhook" references and replace it with admission controller which is a known concept:
https://github.com/kedacore/keda-docs/pull/1027#pullrequestreview-1234204094

File: CHANGELOG.md [hunk 17]
```diff
@@ -49,6 +49,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 Here is an overview of all **stable** additions:
 
 - **General**: Introduce new ArangoDB Scaler ([#4000](https://github.com/kedacore/keda/issues/4000))
+- **General**: Validate incoming ScaledObjects to ensure the workload isn't already autoscaled ([#3755](https://github.com/kedacore/keda/issues/3755))
```
Comment (tomkerkhove) [inline CHANGELOG.md:52]:
```suggestion
- **General**: Introduce admission controller to automatically validate resource changes to prevent misconfiguration and enforce best practices. ([#3755](https://github.com/kedacore/keda/issues/3755))
  - For example, prevent multiple ScaledObjects/HPAs to scale the same scale target
```

File: cmd/adapter/main.go [hunk 18]
```diff
@@ -200,14 +200,7 @@ func runScaledObjectController(ctx context.Context, mgr manager.Manager, scaleHa
 // generateDefaultMetricsServiceAddr generates default Metrics Service gRPC Server address based on the current Namespace.
 // By default the Metrics Service gRPC Server runs in the same namespace on the keda-operator pod.
 func generateDefaultMetricsServiceAddr() string {
-	const defaultNamespace = "keda"
```
Comment (tomkerkhove) [inline cmd/adapter/main.go:203]: Why are we removing this?

File: config/manager/manager.yaml [hunk 19]
```diff
@@ -63,6 +63,7 @@ spec:
             - name: KEDA_HTTP_DEFAULT_TIMEOUT
               value: ""
           securityContext:
+            runAsNonRoot: true
```
Comment (tomkerkhove) [inline config/manager/manager.yaml:66]: I'm not sure if we should sneak this in this PR. Might be better to do in another PR + changelog entry?

File: config/metrics-server/deployment.yaml [hunk 20]
```diff
@@ -72,6 +72,7 @@ spec:
           - mountPath: /apiserver.local.config/certificates
             name: certs
           securityContext:
+            runAsNonRoot: true
```
Comment (tomkerkhove) [inline config/metrics-server/deployment.yaml:75]: I'm not sure if we should sneak this in this PR. Might be better to do in another PR + changelog entry?

File: config/rbac/role.yaml (hunk 12 above)
Comment (tomkerkhove) [reply config/rbac/role.yaml:112]: I agree on splitting because this permission is added to the `keda-operator` role which is another component and this another identity.

File: config/webhooks/kustomization.yaml [hunk 21]
```diff
@@ -0,0 +1,12 @@
...
+- webhooks.yaml
+- secret.yaml
+- service.yaml
+- validation_webhooks.yaml
+
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: ghcr.io/kedacore/keda-webhooks
+  newName: ghcr.io/kedacore/keda-webhooks
```
Comment (tomkerkhove) [inline config/webhooks/kustomization.yaml:11]:
```suggestion
- name: ghcr.io/kedacore/keda-admission-controller
  newName: ghcr.io/kedacore/keda-admission-controller
```

File: config/webhooks/service.yaml [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+kind: Service
+metadata:
+  labels:
+    app.kubernetes.io/name: service
+    app.kubernetes.io/instance: webhook-service
+    app.kubernetes.io/component: webhook
+    app.kubernetes.io/created-by: keda
+    app.kubernetes.io/part-of: keda
+    app.kubernetes.io/managed-by: kustomize
+  name: keda-webhooks
```
Comment (tomkerkhove) [inline config/webhooks/service.yaml:12]:
```suggestion
  name: keda-admission-controller
```

Comment (tomkerkhove) [inline config/webhooks/service.yaml:12]:
```suggestion
    app.kubernetes.io/name: service
    app.kubernetes.io/instance: admission-controller-service
    app.kubernetes.io/component: admission-controller
    app.kubernetes.io/created-by: keda
    app.kubernetes.io/part-of: keda
    app.kubernetes.io/managed-by: kustomize
  name: keda-admission-controller
```

File: cmd/adapter/main.go (hunk 18 above)
Comment (tomkerkhove) [reply cmd/adapter/main.go:203]: Ok, I did not see the inlined method call in sprintf which is in pkg/util/env_resolver.go

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
I'm not a Go expert so I just did a skim but I would remove all "webhook" references and rename it to "admission controller" which is what it actually is.

Also, I would split the RBAC roles so that our admission controller has it's own roles and identity.

File: config/manager/manager.yaml (hunk 19 above)
Comment (JorTurFer) [reply config/manager/manager.yaml:66]: this is just a minor change because the policy is already defined at pod level, I added it to solve a semgrep issue that requires is also at container level

File: config/metrics-server/deployment.yaml (hunk 20 above)
Comment (JorTurFer) [reply config/metrics-server/deployment.yaml:75]: this is just a minor change because the policy is already defined at pod level, I added it to solve a semgrep issue that requires is also at container level

File: config/rbac/role.yaml (hunk 12 above)
Comment (JorTurFer) [reply config/rbac/role.yaml:112]: The question is, do we want the extra work of having to manage 3 different makefile targets for generating 3 different rbac files based on 3 different group of files (most probably, splitting the api files too)? I can give a try, but `kubebuilder` AFAIK doesn't support something like setting an arbitrary role name for each permission you define, it takes all the permissions and creates a single ClusterRole and Role with all of them.
Other option is to split it manually, but I don't like it at all because the CRDs and RBAC manifests are automatically generated

**Comment by JorTurFer** [comment]
> I'm not a Go expert so I just did a skim but I would remove all "webhook" references and rename it to "admission controller" which is what it actually is.

I'm not 100% sure about this in the code because the technical name is webhook, validating and mutating webhooks. I agree with renaming it externally, but internally doesn't make sense IMO.

**Comment by zroubalik** [comment]
> > I'm not a Go expert so I just did a skim but I would remove all "webhook" references and rename it to "admission controller" which is what it actually is.
> 
> I'm not 100% sure about this in the code because the technical name is webhook, validating and mutating webhooks. I agree with renaming it externally, but internally doesn't make sense IMO.

I agree, I'd keep webhook.

**Comment by tomkerkhove** [comment]
But, why call the same thing differently if they are conceptually the same? Also, just calling it webhook is very vague and you might think this is for sending webhooks as KEDA.

**Comment by tomkerkhove** [comment]
Then at least call this `prevention-webhook`, or `admission-controller-webhook` or the like.

The container image should definately be called `keda-admission-controller` imo.

**Comment by JorTurFer** [comment]
> But, why call the same thing differently if they are conceptually the same? Also, just calling it webhook is very vague and you might think this is for sending webhooks as KEDA.

Because end-users naming not always matches with developer matching, I know it's a problem but for new contributors, looking for information about `admission-controller` is more complicated than looking for `validating/mutating webhooks` because admission controller is not the official name https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/

**Comment by JorTurFer** [comment]
> The container image should definately be called `keda-admission-controller` imo.

I agree with name it admission-controller externally, in docs, docker tags, etc. But in the code I'd keep it as it is

**Comment by tomkerkhove** [comment]
> Also, just calling it webhook is very vague and you might think this is for sending webhooks as KEDA.

As mentioned above, webhooks is vague; let's be more explicit then.

Some suggestions:

> Then at least call this `prevention-webhook`, or `admission-controller-webhook` or the like.

**Comment by tomkerkhove** [comment]
They call them `admissionWebhooks`; our codebase calls them `webhooks`. I'm fine with the former but not the latter.

**Comment by tomkerkhove** [comment]
For example:
image

image

image

**Comment by zroubalik** [comment]
@tomkerkhove what is the problem with `webhook` ? I prefer this, as it is widely used in the tech and when developing operators.

**Comment by JorTurFer** [comment]
I meant to change the name externally, admission-webhooks for end-users, webhooks for developers

**Comment by tomkerkhove** [comment]
> But, why call the same thing differently if they are conceptually the same? Also, just calling it webhook is very vague and you might think this is for sending webhooks as KEDA.

This is why I don't like `webhooks` as a name, @zroubalik. It's not explicit enough what it means imo and contributors will be confused.

**Comment by tomkerkhove** [comment]
> I meant to change the name externally, `admission-webhooks` for end-users, `webhooks` for developers. It's what I was doing, indeed. The[ Operator SDK](https://sdk.operatorframework.io/docs/building-operators/golang/webhook/) names them admission webhooks externally and [`webhooks`](https://github.com/kedacore/keda/pull/4001/files#diff-d1fb402f1269ba7396dca9cbd03b44ac055874669bce1d0f239fa2f54ea3584bR17) in the generated resources

Externally it should be called admission controller but I guess it must be me here.

**Comment by zroubalik** [comment]
> > I meant to change the name externally, `admission-webhooks` for end-users, `webhooks` for developers. It's what I was doing, indeed. The[ Operator SDK](https://sdk.operatorframework.io/docs/building-operators/golang/webhook/) names them admission webhooks externally and [`webhooks`](https://github.com/kedacore/keda/pull/4001/files#diff-d1fb402f1269ba7396dca9cbd03b44ac055874669bce1d0f239fa2f54ea3584bR17) in the generated resources
> 
> Externally it should be called admission controller but I guess it must be me here.

controller and webhook are two different things, admission controller is confusing imoh (I know it is being used), but this stuff is a webhook, not a controller. Calling this yet another controller mixes the concepts used in the codebase already.

**Comment by JorTurFer** [comment]
> controller and webhook are two different things, admission controller is confusing imho (I know it naming is being used), but this stuff is a webhook, not a controller. Calling this yet another controller mixes the concepts used in the codebase already.

kubebuilder doesn't call this controller neither and IMO we should follow the framework rules for development

**Comment by zroubalik** [comment]
> > But, why call the same thing differently if they are conceptually the same? Also, just calling it webhook is very vague and you might think this is for sending webhooks as KEDA.
> 
> This is why I don't like `webhooks` as a name, @zroubalik. It's not explicit enough what it means imo and contributors will be confused.

It's the other way round, kubernetes admission/mutating webhooks are common concetps, controller is something completely different in k8s world. 🤷‍♂️ 

TBH I think we over engineer here again 😄  My vote is to name this a `webhook` everywhere, to avoid complicating things.

**Comment by tomkerkhove** [comment]
:shrug: I'll stop bothering and let you move on.

**Comment by zroubalik** [comment]
> 🤷 I'll stop bothering and let you move on.

Full of energy after the holidays, right? 😄  :trollface:

File: cmd/webhooks/main.go (hunk 6 above)
Comment (JorTurFer) [reply cmd/webhooks/main.go]: I have renamed the parameter to `enable-cert-rotation`. This parameter with ensure that the secret provided by `webhooks-cert-secret-name` exists and will place the certificates there. 
For scenarios with user-provided certificates, this parameter has to be removed or switched to false and also users have to ensure that valid certificates are placed in the folder given by `webhooks-cert-dir` (`/certs` as default).
I have created [an issue](https://github.com/kedacore/keda-docs/issues/1030) in docs to properly documented because I'd like to document also other certificates, all together in a specific section

File: config/rbac/role.yaml (hunk 12 above)
Comment (JorTurFer) [reply config/rbac/role.yaml]: I have split this and now webhooks component installation adds the extra permissions

**Comment by JorTurFer** [comment]
PTAL @tomkerkhove @zroubalik .
I have done (from your last review):
- RBAC files are split
- 2 yamls are generated and pushed to release, keda-xxxx.yaml and keda-xxxx-core.yaml with and without webhooks
- Renamed the parameter to enable cert-rotation
- Updated the docs to say that where are the webhooks to starting contributions
- Added the CPU/Memory validation **for deployments and statefulsets only** because CRDs can be anything no having resources section
- Renamed external name to admission-webhooks

I have to update docs yet, but I think this PR is ready

**Comment by JorTurFer** [comment]
I'm still thinking about this, and I have doubts about using the operator for cert management or if it could be worth to introduce another component just for managing the certificates and the ca injections (like https://github.com/jet/kube-webhook-certgen/ does, but generating a bundle cert with all our requirements, not just 1 cert per thing). The architecture grows from 2 components to 4, but the certificates component is the piece that end users have to replace with cert-manager or other tool. Or maybe I'm overthinking, and we can introduce this new component in the future if it's necessary.

File: config/manager/manager.yaml [hunk 23]
```diff
@@ -35,6 +35,7 @@ spec:
             - --zap-log-level=info
             - --zap-encoder=console
             - --zap-time-encoding=rfc3339
+            - --enable-cert-rotation=true
```
Comment (tomkerkhove) [inline config/manager/manager.yaml:38]: Is this another feature added in this PR? If so, let's update changelog

File: config/rbac/role.yaml [hunk 24]
```diff
@@ -43,6 +43,16 @@ rules:
...
+- apiGroups:
+  - admissionregistration.k8s.io
+  resources:
+  - validatingwebhookconfigurations
+  verbs:
+  - get
+  - list
+  - patch
+  - update
+  - watch
```
Comment (tomkerkhove) [inline config/rbac/role.yaml:55]: Does the operator still need this? I don't think so, right?

File: config/webhooks/webhooks.yaml [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+            - --zap-encoder=console
+            - --zap-time-encoding=rfc3339
+          imagePullPolicy: Always
+          resources:
+            requests:
+              cpu: 100m
+              memory: 100Mi
+            limits:
+              cpu: 1000m
+              memory: 1000Mi
```
Comment (tomkerkhove) [inline config/webhooks/webhooks.yaml:44]: Let's update our doc PR to include this on https://keda.sh/docs/2.9/operate/cluster/#cluster-capacity, but for 2.10

File: config/rbac/role.yaml (hunk 24 above)
Comment (JorTurFer) [reply config/rbac/role.yaml:55]: yeah, the operator is now who generates the certificates and also patches the k8s manifests with the caBundle (validatingwebhookconfigurations in this PR and in the nexts also apiservices), the admission webhooks doesn't require any extra permission.

File: config/manager/manager.yaml (hunk 23 above)
Comment (JorTurFer) [reply config/manager/manager.yaml:38]: No, this is part of the cert generation, moved from the webhooks to the operator for the incoming features, but this is required for this feature and used only for it atm

File: config/webhooks/webhooks.yaml (hunk 25 above)
Comment (JorTurFer) [reply config/webhooks/webhooks.yaml:44]: sure, omw

File: CHANGELOG.md [hunk 26]
```diff
@@ -49,6 +49,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 Here is an overview of all **stable** additions:
 
 - **General**: Introduce new ArangoDB Scaler ([#4000](https://github.com/kedacore/keda/issues/4000))
+- **General**: Introduce admission webhooks to automatically validate resource changes to prevent misconfiguration and enforce best practices. ([#3755](https://github.com/kedacore/keda/issues/3755))
```
Comment (zroubalik) [inline CHANGELOG.md:52]: nit, please sort alphabetically

File: Makefile [hunk 27]
```diff
@@ -116,7 +117,7 @@ smoke-test: ## Run e2e tests against Kubernetes cluster configured in ~/.kube/co
 
 ##@ Development
 
-manifests: controller-gen ## Generate ClusterRole and CustomResourceDefinition objects.
+manifests:  controller-gen ## Generate ClusterRole and CustomResourceDefinition objects.
```
Comment (zroubalik) [inline Makefile:120]: why the extra space here?

File: cmd/operator/main.go [hunk 28]
```diff
@@ -256,3 +280,101 @@ func main() {
 		os.Exit(1)
 	}
 }
+
+// +kubebuilder:rbac:groups=admissionregistration.k8s.io,resources=validatingwebhookconfigurations,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",namespace=keda,resources=secrets,verbs=get;list;watch;create;update;patch;delete
```
Comment (zroubalik) [inline cmd/operator/main.go:285]: this should be probably deleted, right?

File: cmd/operator/main.go [hunk 29]
```diff
@@ -256,3 +280,101 @@ func main() {
...
+
+// addCertificateRotation registers all needed services to generate the certificates and patches needed resources with the caBundle
+func addCertificateRotation(ctx context.Context, mgr manager.Manager, secretName, certDir, operatorServiceName, metricsServerServiceName, webhooksServiceName string) {
```
Comment (zroubalik) [inline cmd/operator/main.go:288]: Would be nice to move all certificate related stuff to a separate package/file, but this doesn't have to be done in this PR. But in the following one, dealing with the global certs solution.

File: cmd/operator/main.go [hunk 30]
```diff
@@ -256,3 +280,101 @@ func main() {
...
+		CAOrganization:         caOrganization,
+		DNSName:                extraDNSNames[0],
+		ExtraDNSNames:          extraDNSNames,
+		IsReady:                setupFinished,
+		Webhooks:               rotatorHooks,
+		RestartOnSecretRefresh: true,
+		RequireLeaderElection:  true,
+	}); err != nil {
+		setupLog.Error(err, "unable to set up cert rotation")
+		os.Exit(1)
```
Comment (zroubalik) [inline cmd/operator/main.go:324]: this function should return err and exit in the main

File: cmd/operator/main.go [hunk 31]
```diff
@@ -256,3 +280,101 @@ func main() {
...
+					"app.kubernetes.io/name":      "keda-operator",
+					"app.kubernetes.io/component": "keda-operator",
+					"app.kubernetes.io/part-of":   "keda",
+				},
+			},
+		}
+		err = mgr.GetClient().Create(ctx, secret)
+		if err != nil {
+			setupLog.Error(err, "unable to create certificates secret")
+			os.Exit(1)
```
Comment (zroubalik) [inline cmd/operator/main.go:376]: return err and exit in the main

**Review by zroubalik — COMMENTED** [review]
Looking really great, just a minor nits.

File: config/manager/manager.yaml (hunk 23 above)
Comment (zroubalik) [reply config/manager/manager.yaml:38]: I'd keep it without the external component.

File: cmd/operator/main.go (hunk 28 above)
Comment (JorTurFer) [reply cmd/operator/main.go:285]: no, this is required in the operator because it generates the certificates and stores them as secret in keda namespace. I plan to deploy this conditionally in helm for cases where end-users manage the certs externally

File: Makefile (hunk 27 above)
Comment (JorTurFer) [reply Makefile:120]: typo

Comment (JorTurFer) [reply Makefile:120]:
```suggestion
manifests: controller-gen ## Generate ClusterRole and CustomResourceDefinition objects.
```

File: CHANGELOG.md (hunk 26 above)
Comment (JorTurFer) [reply CHANGELOG.md:52]: sure!

File: cmd/operator/main.go (hunk 29 above)
Comment (JorTurFer) [reply cmd/operator/main.go:288]: I plan to refactor the certs in operator and ms after this, I'll address this change there

File: cmd/operator/main.go (hunk 31 above)
Comment (JorTurFer) [reply cmd/operator/main.go:376]: nice catch! I moved this code and I forgot the exit part

**Comment by JorTurFer** [comment]
Ready for the final review :)

File: cmd/operator/main.go (hunk 28 above)
Comment (zroubalik) [reply cmd/operator/main.go:288]: and this it have to be here? wouldn't be webhook a better location?

Comment (JorTurFer) [reply cmd/operator/main.go:288]: no, because it's the operator who does the job xD I mean, even a user doesn't use the webhooks, the operator still generates the certificates (because they will be used for the other communications). The operator needs access to secrets in keda namespace

Comment (zroubalik) [reply cmd/operator/main.go:288]: What I am trying to say is that these markers a little bit "lost" in this file.

Comment (JorTurFer) [reply cmd/operator/main.go:288]: oh, good point! I have placed it there because it is closest to the code where it's needed, when I move this to its own package I'll move them there also

**Review by zroubalik — COMMENTED** [review]
/run-e2e internals*

**Review by zroubalik — APPROVED** [review]
LGTM, great job!

File: cmd/operator/main.go (hunk 28 above)
Comment (zroubalik) [reply cmd/operator/main.go:288]: yeah :)

**Comment by JorTurFer** [comment]
Is it okey for you @tomkerkhove ? can I merge the feature?

**Review by tomkerkhove — APPROVED** [review]
A small approve for a review, a big step for KEDA
